### PR TITLE
[queue arch] remove base-runner deps.

### PR DESCRIPTION
**Pull request by zchcai**
Separate https://github.com/zchcai/fuzzbench/pull/10/files with two parts:
1. the docker build changes (the elimination of the double dependency) in this PR
- Delete docker/base-runner
- Update the image_types.yaml
- Modify docker/gcb yaml files for consistency
2. the modification of the run_experiment/worker python code in #585. 
- Modify docker/generate_makefile.py to get the change tested with queue-based arch CI.

This PR is blocked by #584, where oss-fuzz benchmarks are the same functionality for normal fuzzbench.
Then we try to get a mono-dependency build logic, i.e. remove multi-dependency jobs.
At last, we merge #585 for an integration test.

Generally, #550 will be my real experiment environment, but not suitable for directly merge yet.

### CODE REVIEW

**Comment by zchcai** [comment]
Since many benchmarks will use `base-runner`, if we move the apt-install in a future PR, it may broke the process since benchmark writers will have a different base images for usage.

**Comment by zchcai** [comment]
The CI test will fail because we delete the base-runner in the dependency tree.
But our fuzzbench/run_experiment will build base-runner in advance. Thus, an not found errer shows up.

**Comment by lszekeres** [comment]
Could you update the Makefile (and GCB yaml-s) too to get the change tested by CI?

**Comment by zchcai** [comment]
> Could you update the Makefile (and GCB yaml-s) too to get the change tested by CI?

Good suggestion to try. My concern is, this is not an issue can be resolved by changing Makefile or GCB yamls. This is the dependency for all fuzzer's runner.Dockerfile. See https://github.com/google/fuzzbench/pull/601#issuecomment-665119261

**Comment by zchcai** [comment]
> > Could you update the Makefile (and GCB yaml-s) too to get the change tested by CI?
> 
> Good suggestion to try. My concern is, this is not an issue can be resolved by changing Makefile or GCB yamls. This is the dependency for all fuzzer's runner.Dockerfile. See [#601 (comment)](https://github.com/google/fuzzbench/pull/601#issuecomment-665119261)

If we directly remove base-runner, for example, put base-image here, how to make sure the logic is still valid for these runner.Dockerfile?
```
aflplusplus_laf_cmplog/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

afl_qemu/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

aflplusplus_qemu/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

fastcgs_lm/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

honggfuzz/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

honggfuzz_qemu/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

aflplusplus_optimal/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

aflcc/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

eclipser/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

aflfast/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

afl/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

aflplusplus_havoc2/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

aflplusplus_havoc_cmplog/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

aflplusplus/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

aflplusplus_cmplog/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

aflplusplus_honggfuzz/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

libfuzzer/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

aflplusplus_havoc/runner.Dockerfile
16:FROM gcr.io/fuzzbench/base-runner

aflplusplus_laf/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

mopt/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

manul/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

entropic/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

lafintel/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

aflsmart/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

fairfuzz/runner.Dockerfile
15:FROM gcr.io/fuzzbench/base-runner

**Comment by lszekeres** [comment]
> > > Could you update the Makefile (and GCB yaml-s) too to get the change tested by CI?
> > 
> > 
> > Good suggestion to try. My concern is, this is not an issue can be resolved by changing Makefile or GCB yamls. This is the dependency for all fuzzer's runner.Dockerfile. See [#601 (comment)](https://github.com/google/fuzzbench/pull/601#issuecomment-665119261)
> 
> If we directly remove base-runner, for example, put base-image here, how to make sure the logic is still valid for these runner.Dockerfile?
> 
> ```
> aflplusplus_laf_cmplog/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> afl_qemu/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> aflplusplus_qemu/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> fastcgs_lm/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> honggfuzz/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> honggfuzz_qemu/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> aflplusplus_optimal/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> aflcc/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> eclipser/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> aflfast/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> afl/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> aflplusplus_havoc2/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> aflplusplus_havoc_cmplog/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> aflplusplus/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> aflplusplus_cmplog/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> aflplusplus_honggfuzz/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> libfuzzer/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> aflplusplus_havoc/runner.Dockerfile
> 16:FROM gcr.io/fuzzbench/base-runner
> 
> aflplusplus_laf/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> mopt/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> manul/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> entropic/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> lafintel/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> aflsmart/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> 
> fairfuzz/runner.Dockerfile
> 15:FROM gcr.io/fuzzbench/base-runner
> ```

Right, as part of eliminating `base-runner` we need to change those `FROM`-s from `base-runner` to `base-image` -- as we are moving what happened in `base-runner` into `benchmark-builder`. This PR should also delete `docker/base-runner` so we can make sure nothing relies on it and the CI passes.

**Comment by zchcai** [comment]
> Right, as part of eliminating base-runner we need to change those FROM-s from base-runner to base-image -- as we are moving what happened in base-runner into benchmark-builder. This PR should also delete docker/base-runner so we can make sure nothing relies on it and the CI passes.

OK. Then we focus on the CI tests in old arch for this PR. And we resolve the new arch issues in #550.

**Comment by zchcai** [comment]
Seems benchmark `proj4-2017-08-14` has some problems.
```
subprocess.CalledProcessError: Command '['/bin/bash', '-ex', '/src/build.sh']' returned non-zero exit status 8.
The command '/bin/sh -c PYTHONPATH=$SRC python3 -u -c "from fuzzers import utils; utils.initialize_flags(); import fuzzer; fuzzer.build()"' returned a non-zero code: 1
make: *** [.aflsmart-proj4-2017-08-14-builder] Error 1
docker/generated.mk:19164: recipe for target '.aflsmart-proj4-2017-08-14-builder' failed
Building benchmarks: libjpeg-turbo-07-2017, freetype2-2017, proj4-2017-08-14, libpng-1.2.56, re2-2014-12-09, libxml2-v2.9.2, lcms-2017-03-21, woff2-2016-05-06, openthread-2019-12-23, perl-5.21.7, harfbuzz-1.3.2, vorbis-2017-12-11 for fuzzer: aflsmart
Running command: make RUNNING_ON_CI=yes -j test-run-aflsmart-libjpeg-turbo-07-2017
Running command: make RUNNING_ON_CI=yes -j test-run-aflsmart-freetype2-2017
Running command: make RUNNING_ON_CI=yes -j test-run-aflsmart-proj4-2017-08-14
##[error]Process completed with exit code 1.

**Comment by zchcai** [comment]
> Seems benchmark `proj4-2017-08-14` has some problems.
> 
> ```
> subprocess.CalledProcessError: Command '['/bin/bash', '-ex', '/src/build.sh']' returned non-zero exit status 8.
> The command '/bin/sh -c PYTHONPATH=$SRC python3 -u -c "from fuzzers import utils; utils.initialize_flags(); import fuzzer; fuzzer.build()"' returned a non-zero code: 1
> make: *** [.aflsmart-proj4-2017-08-14-builder] Error 1
> docker/generated.mk:19164: recipe for target '.aflsmart-proj4-2017-08-14-builder' failed
> Building benchmarks: libjpeg-turbo-07-2017, freetype2-2017, proj4-2017-08-14, libpng-1.2.56, re2-2014-12-09, libxml2-v2.9.2, lcms-2017-03-21, woff2-2016-05-06, openthread-2019-12-23, perl-5.21.7, harfbuzz-1.3.2, vorbis-2017-12-11 for fuzzer: aflsmart
> Running command: make RUNNING_ON_CI=yes -j test-run-aflsmart-libjpeg-turbo-07-2017
> Running command: make RUNNING_ON_CI=yes -j test-run-aflsmart-freetype2-2017
> Running command: make RUNNING_ON_CI=yes -j test-run-aflsmart-proj4-2017-08-14
> ##[error]Process completed with exit code 1.
> ```

However, if I run `make test-run-aflsmart-proj4-2017-08-14` locally, it successes.

**Comment by zchcai** [comment]
Also, the reported error didn't show in local test:
```
make test-run-aflplusplus_havoc_cmplog-libpng-1.2.56
...
INFO:root:Doing trial as part of experiment.
INFO:root:Starting trial.
[run_afl_fuzz] Running target with afl-fuzz
[run_afl_fuzz] Running command: ./afl-fuzz -i /out/seeds -o /out/corpus -d -m none -c /out/cmplog/fuzz-target -- /out/fuzz-target 2147483647
[+] Loaded environment variable AFL_NO_UI with value 1
[+] Loaded environment variable AFL_SKIP_CPUFREQ with value 1
[+] Loaded environment variable AFL_NO_AFFINITY with value 1
[+] Loaded environment variable AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES with value 1
afl-fuzz++2.66d based on afl by Michal Zalewski and a big online community
[+] afl++ is maintained by Marc "van Hauser" Heuse, Heiko "hexcoder" Eißfeldt, Andrea Fioraldi and Dominik Maier
[+] afl++ is open source, get it at https://github.com/AFLplusplus/AFLplusplus
[+] Power schedules from github.com/mboehme/aflfast
[+] Python Mutator and llvm_mode instrument file list from github.com/choller/afl
[+] MOpt Mutator from github.com/puppet-meteor/MOpt-AFL
[*] Getting to work...
[+] Using exploration-based constant power schedule (EXPLORE, default)
[+] Disabling the UI because AFL_NO_UI is set.
[*] Checking core_pattern...
[+] You have 16 CPU cores and 1 runnable tasks (utilization: 6%).
[+] Try parallel jobs - see docs/parallel_fuzzing.md.
[*] Setting up output directories...
[+] Output directory exists but deemed OK to reuse.
[*] Deleting old session data...
[+] Output dir cleanup successful.
[!] WARNING: Not binding to a CPU core (AFL_NO_AFFINITY set).
[*] Scanning '/out/seeds'...
[+] No auto-generated dictionary tokens to reuse.
[*] Creating hard links for all input files...
[*] Validating target binary...
[+] Persistent mode binary detected.
[+] Deferred forkserver binary detected.
[*] Validating target binary...
[+] Persistent mode binary detected.
[+] Deferred forkserver binary detected.
[*] Spawning cmplog forkserver
[*] Spinning up the fork server...
[+] All right - fork server is up.
[*] Using SHARED MEMORY FUZZING feature.
[*] Target map size: 3008
[+] Cmplog forkserver successfully started
[*] Attempting dry run with 'id:000000,time:0,orig:1d235a0efef192095e9a4de48a1f1960a2cae3ab'...
[*] Spinning up the fork server...
[+] All right - fork server is up.
[*] Using SHARED MEMORY FUZZING feature.
[*] Target map size: 2984
    len = 1260, map size = 235, exec speed = 220 us
[+] All test cases processed.

[+] Here are some useful stats:

    Test case count : 1 favored, 0 variable, 1 total
       Bitmap range : 235 to 235 bits (average: 235.00 bits)
        Exec timing : 220 to 220 us (average: 220 us)

[*] No -t option specified, so I'll use exec timeout of 20 ms.
[+] All set and ready to roll!
[*] Entering queue cycle 1.
[*] Fuzzing test case #0 (1 total, 0 uniq crashes found)...
[*] Fuzzing test case #1 (101 total, 0 uniq crashes found)...
[*] Fuzzing test case #3 (102 total, 0 uniq crashes found)...
[*] Fuzzing test case #8 (102 total, 0 uniq crashes found)...
[*] Fuzzing test case #9 (110 total, 0 uniq crashes found)...
[*] Fuzzing test case #11 (110 total, 0 uniq crashes found)...
[*] Fuzzing test case #21 (119 total, 0 uniq crashes found)...
[*] Fuzzing test case #28 (120 total, 0 uniq crashes found)...
[*] Fuzzing test case #35 (124 total, 0 uniq crashes found)...
[*] Fuzzing test case #39 (125 total, 0 uniq crashes found)...
[*] Fuzzing test case #40 (126 total, 0 uniq crashes found)...
[*] Fuzzing test case #41 (126 total, 0 uniq crashes found)...
[*] Fuzzing test case #42 (128 total, 0 uniq crashes found)...
[*] Fuzzing test case #44 (128 total, 0 uniq crashes found)...
[*] Fuzzing test case #47 (129 total, 0 uniq crashes found)...
[*] Fuzzing test case #48 (129 total, 0 uniq crashes found)...
[*] Fuzzing test case #49 (129 total, 0 uniq crashes found)...
[*] Fuzzing test case #50 (129 total, 0 uniq crashes found)...
[*] Fuzzing test case #51 (129 total, 0 uniq crashes found)...
[*] Fuzzing test case #52 (129 total, 0 uniq crashes found)...
[*] Fuzzing test case #53 (129 total, 0 uniq crashes found)...
[*] Fuzzing test case #54 (129 total, 0 uniq crashes found)...
INFO:root:Doing final sync.
```

File: docker/gcb/base-images.yaml [hunk 1]
```diff
@@ -53,36 +52,6 @@ steps:
...
-      docker pull ${_REPO}/base-runner || exit 0
-  id: 'base-runner-pull'
-  wait_for: ['-']  # Start this immediately.
-
-- name: 'gcr.io/cloud-builders/docker'
-  args: [
-    'build',
-
-    '--tag',
-    'gcr.io/fuzzbench/base-builder',
```
Comment (zchcai) [inline docker/gcb/base-images.yaml:71]: By the way, before deleting, this line should be `base-runner` I guess.

**Comment by zchcai** [comment]
Actually, local run for `make test-run-fastcgs_lm-sqlite3_ossfuzz` works, no errors.

**Comment by lszekeres** [comment]
looks like failures are due to network / download issues ? @jonathanmetzman have you seen these before

```
---> 178b709312ea
Step 3/6 : RUN mkdir $SRC/sqlite3 &&     cd $SRC/sqlite3 &&     curl 'https://sqlite.org/src/tarball/sqlite.tar.gz?r=c78cbf2e86850cc6' -o sqlite3.tar.gz &&         tar xzf sqlite3.tar.gz --strip-components 1
 ---> Running in e0b6f0a67bde
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  5278  100  5278    0     0  12007      0 --:--:-- --:--:-- --:--:-- 12022

gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now
```

**Comment by zchcai** [comment]
> LGTM if CI passes

We only need to watch whether this one will pass for a rerun.
```
37 successful and 1 failing checks
Build fuzzers / build (fastcgs_lm, oss-fuzz) (pull_request) Failing after 47m

**Comment by zchcai** [comment]
> > LGTM if CI passes
> 
> We only need to watch whether this one will pass for a rerun.
> 
> ```
> 37 successful and 1 failing checks
> Build fuzzers / build (fastcgs_lm, oss-fuzz) (pull_request) Failing after 47m
> ```

Let's wait for [this](https://github.com/google/fuzzbench/pull/601/checks?check_run_id=925869123).

**Comment by zchcai** [comment]
> > > LGTM if CI passes
> > 
> > 
> > We only need to watch whether this one will pass for a rerun.
> > ```
> > 37 successful and 1 failing checks
> > Build fuzzers / build (fastcgs_lm, oss-fuzz) (pull_request) Failing after 47m
> > ```
> 
> Let's wait for [this](https://github.com/google/fuzzbench/pull/601/checks?check_run_id=925869123).

It passes.
```
Build fuzzers / build (fastcgs_lm, oss-fuzz)
succeeded 12 hours ago in 1h 11m 1s

**Comment by zchcai** [comment]
Just to make sure: there will be no base-builder and no base-runner after this PR, right?

File: docker/generate_makefile.py [hunk 2]
```diff
@@ -25,14 +25,14 @@
 """
 
 FUZZER_TEMPLATE = """
-.{fuzzer}-builder: base-builder
+.{fuzzer}-builder: base-image
```
Comment (zchcai) [inline docker/generate_makefile.py:28]: Here, I change this to base-image

**Comment by zchcai** [comment]
need to wait for the only one:
Build fuzzers / build (eclipser, standard) (pull_request) Cancelled after 360m
Then we can merge.
Now I rerun it.

**Comment by zchcai** [comment]
> need to wait for the only one:
> Build fuzzers / build (eclipser, standard) (pull_request) Cancelled after 360m
> Then we can merge.
> Now I rerun it.

It passes.

**Pull request merged**